### PR TITLE
changeset: add struct typing to changeset

### PR DESCRIPTION
**Pull request by floodfx**

### CODE REVIEW

File: changeset/funcs.go [hunk 1]
```diff
@@ -6,16 +6,6 @@ import (
 	"github.com/canopyclimate/golive/htmltmpl"
 )
 
-// Funcs returns a map of functions that can be used in templates:
-//   - inputTag: renders an input tag with the given key and value for the provided changeset
-//   - errorTag: renders an error tag if there is an error for the given key in the provided changeset
-func Funcs() htmltmpl.FuncMap {
```
Comment (kevboh) [inline changeset/funcs.go:12]: Why did these go away?

File: examples/main.go [hunk 2]
```diff
@@ -237,15 +233,15 @@ func (c *Counter) HandleEvent(ctx context.Context, e *live.Event) error {
...
-	return c, htmltmpl.Must(htmltmpl.New("liveView").Funcs(myFuncs).Parse(`
-			{{ define "inputTag" }}
-				<input type="text" name="{{ .Name }}" value="{{ .Changeset.Value .Name }}"/>
-			{{ end }}
-			{{ define "errorTag" }}
-				{{ if .Changeset.HasError .Name}}
-					<span class="error">{{ .Changeset.Error .Name }}</span>
-				{{ end }}
-			{{ end }}
+	funcs := htmltmpl.FuncMap{
```
Comment (kevboh) [inline examples/main.go:236]: Oh, I assume because we don't know the generic type until runtime

**Review by kevboh — APPROVED** [review]
Much cleaner! I don't trust myself to fully understand the repercussions of generics here, so take my approval with a grain of salt

File: changeset/funcs.go (hunk 1 above)
Comment (floodfx) [reply changeset/funcs.go:12]: Only logical thing we could type these functions with is `any` but that causes a runtime error unless the `Changeset` type is `any` which is obviously not helpful.

File: examples/main.go (hunk 2 above)
Comment (floodfx) [reply examples/main.go:236]: this seemed like the only way to add typed `Changeset` specific funcs without resorting to some naming convention that explodes with the number of typed `Changesets`.  LMK if you have any better ideas.

File: changeset/funcs.go (hunk 1 above)
Comment (josharian) [reply changeset/funcs.go:12]: See my suggestion below...

File: changeset/funcs.go [hunk 3]
```diff
@@ -26,7 +16,7 @@ var (
 )
 
 // InputTag renders an input tag with the given key and value for the provided changeset.
-func InputTag(cs *Changeset, key string) htmltmpl.HTML {
+func InputTag[T any](cs *Changeset[T], key string) htmltmpl.HTML {
```
Comment (josharian) [inline changeset/funcs.go:19]: How about instead something like:

```go
type Valuer interface {
  Value(string)string
}

func InputTag(v Valuer, key string) htmltmpl.HTML {
```

**Review by josharian — COMMENTED** [review]
this is gonna be nice!

**Pull request merged**